src/ir/proto/predicate.cc

```diff
@@ -0,0 +1,94 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#include "src/ir/proto/predicate.h"
+
+#include <memory>
+
+#include "src/ir/predicates/predicate.h"
+
+namespace raksha::ir::proto {
+
+namespace ir = raksha::ir;
+namespace pred = raksha::ir::predicates;
+
+std::unique_ptr<pred::Predicate> Decode(
+    const arcs::InformationFlowLabelProto_Predicate &predicate_proto) {
+  switch (predicate_proto.predicate_case()) {
+    case arcs::InformationFlowLabelProto_Predicate::kLabel: {
+      const arcs::InformationFlowLabelProto &label = predicate_proto.label();
+      CHECK(label.has_semantic_tag())
+        << "Found a label without required field tag.";
+      return std::make_unique<pred::TagPresence>(label.semantic_tag());
+    }
+    case arcs::InformationFlowLabelProto_Predicate::kAnd: {
+      const arcs::InformationFlowLabelProto_Predicate_And &and_predicate =
+          predicate_proto.and_();
+      CHECK(and_predicate.has_conjunct0())
+        << "Found an And predicate without required field conjunct0.";
+      CHECK(and_predicate.has_conjunct1())
+        << "Found an And predicate without required field conjunct1.";
+      std::unique_ptr<pred::Predicate> conjunct0_predicate =
+          Decode(and_predicate.conjunct0());
+      std::unique_ptr<pred::Predicate> conjunct1_predicate =
+          Decode(and_predicate.conjunct1());
+      return std::make_unique<pred::And>(
+          std::move(conjunct0_predicate), std::move(conjunct1_predicate));
+    }
+    case arcs::InformationFlowLabelProto_Predicate::kImplies: {
+       const arcs::InformationFlowLabelProto_Predicate_Implies
+        &implies_predicate = predicate_proto.implies();
+       CHECK(implies_predicate.has_antecedent())
+        << "Found an Implies predicate without required field antecedent.";
+       CHECK(implies_predicate.has_consequent())
+        << "Found an Implies predicate without required field consequent.";
+       std::unique_ptr<pred::Predicate> antecedent_predicate =
+           Decode(implies_predicate.antecedent());
+       std::unique_ptr<pred::Predicate> consequent_predicate =
+           Decode(implies_predicate.consequent());
+
+       return std::make_unique<pred::Implies>(
+           std::move(antecedent_predicate), std::move(consequent_predicate));
+
+    }
+    case arcs::InformationFlowLabelProto_Predicate::kNot: {
+      const arcs::InformationFlowLabelProto_Predicate_Not &not_predicate =
```

Comment:
`const auto &not_predicate`?

Reply:
See other comment about `auto` use and style guide.